core/src/main/java/org/apache/iceberg/variants/VariantBuilder.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.variants;
+
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonParseException;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.core.exc.InputCoercionException;
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.util.List;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+/** A builder class to build a primitive/array/object variant. */
+public class VariantBuilder extends VariantBuilderBase {
+  public VariantBuilder() {
+    super(new VariantBuilderBase.ByteBufferWrapper(), new VariantBuilderBase.Dictionary());
+  }
+
+  public VariantPrimitiveBuilder createPrimitive() {
+    VariantPrimitiveBuilder primitiveBuilder = new VariantPrimitiveBuilder(getBuffer(), getDict());
+    return primitiveBuilder;
+  }
+
+  public VariantObjectBuilder startObject() {
+    return new VariantObjectBuilder(getBuffer(), getDict());
+  }
+
+  public VariantArrayBuilder startArray() {
+    return new VariantArrayBuilder(getBuffer(), getDict());
+  }
+
+  /**
+   * Parses a JSON string and constructs a Variant object.
+   *
+   * @param json The JSON string to parse.
+   * @return The constructed Variant object.
+   * @throws IOException If an error occurs while reading or parsing the JSON.
+   */
+  public static Variant parseJson(String json) throws IOException {
+    Preconditions.checkArgument(
+        json != null && !json.isEmpty(), "Input JSON string cannot be null or empty.");
+
+    try (JsonParser parser = new JsonFactory().createParser(json)) {
```

Comment:
There are existing utilities in `JSONUtil` for a lot of the boilerplate around parsing and generating JSON. Is there a reason not to use the same approach here?

Reply:
We need to parse the JSON token by token so we can encode them in binary. Seems we only have the version to parse the JSON string into a full JSONNode and then create the object.